proposals/gateway-plusplus-phase1.md

```diff
@@ -0,0 +1,122 @@
+
+HTTP is the lowest common denominator so when we see environmental constraints in application stacks we see HTTP as the means by which developers work around these limitation in order to access IPFS. This also allows developers to leverage additional infrastructure they already have deployed for HTTP, like caching, with IPFS.
+
+We should lean into this rather than push against it. In the short term, we can’t expect every user and developer to have a full implementation of the protocol. Strengthening the Gateway protocol to meet user needs is the path of least resistance and can enable numerous thin clients to be built for IPFS that can run in virtually any environment.
+
+This proposal establishes "Gateway++" as a meta-project for meeting these user needs with extensions to our gateway protocols and infra.
+
+## Problem Statement
+
+For reading data, the IPFS Gateway is already serving these users quite well. Not only does it allow them to read data from the IPFS network without running a full node, they are able to integrate with existing HTTP caching infrastructure to improve performance.
+
+However, the user needs for writing data to the network are not being met well by our current solutions. We have two interfaces that support writes over HTTP and neither was designed as a full solution to these user needs.
+
+- Pinning Service API.
+    - 😡 Requires you to already be running a full IPFS node as a client.
+    - 😡 Is not transactional. You have to poll in order to tell when the data is actually available.
+    - 😃 Is intended to be multi-tenant and authenticated using a flexible bearer token.
+- HTTP RPC API
+    - 😡 Is single-tenant. No authentication system.
+        - The RPC API is designed for a single user to manage a remote node, it was not designed and built to be multi-tenant and would look very different if it were.
+    - 😡 Is not always transactional. Many write calls return before they are available.
+
+These interfaces were built with particular user profiles in mind. The users we’re seeing in NFTs (including our own service as a user of IPFS) are not well served by these interfaces and we need a new “project” definition (Gateway++) that is centered around these user needs.
+
+Developers are building applications that have to run in a lot of constrained environments that can’t support a full node. Strengthening the Gateway to serve these needs over HTTP leverages endpoints they are already using and provides the features they need in a stateless protocol that has ubiquitous support.
+
+We are never going to hear “Sorry, I can’t use HTTP in this environment” from a developer.
+
+## Phase 1
+
+We should expect this to be a long term project with additional proposals in due time.
+
+For now, we have a list of high priority tasks that need to be fixed.
+
+In [nft.storage](http://nft.storage) we have the following high priority needs:
+
+- Add the Pinning API to ipfs-cluster.
```

Comment:
:+1: for adding [Pinning Service API](https://ipfs.github.io/pinning-services-api-spec/) to ipfs-cluster – this will not only help with NFTs, but enable people to self-host pinning infra with ease and use it with [ipfs-webui v2.12.0+](https://github.com/ipfs/ipfs-webui/releases/v2.12.0) and soon ipfs-desktop and Brave.

Reply:
What is the benefit of adding the pinning service API to cluster when cluster already provides the required push API? Is it purely for client code generation and auth tokens?

Reply:
Good point. It would be nice if Cluster did support the pinning service, but adding files and pinning dags to a remote cluster is already well supported. We make use of this in [adding websites to cluster from CI](https://github.com/ipfs-shipyard/ipfs-dns-deploy/blob/7f318679a1c94aeb3b5b92dc8c7bc76b98f39813/scripts/pin-to-cluster.sh#L51-L68) which is a great example of a constrained environment.

Reply:
How much of that can be replicated in a browser if you had the auth available? Is `ipfs-cluster-ctl` just a simple wrapper around the REST API + some UnixFS slurping?

Reply:
yes! `ipfs-cluster-ctl` is a wrapper around the cluster REST api. The auth is basic, so with some https, you're good.

> ipfs-cluster-ctl is an HTTP API client to the REST API endpoint with full feature-parity that always works with the HTTP API as offered by a cluster peer on the same version. Anything that ipfs-cluster-ctl can do is supported by the REST API.

https://cluster.ipfs.io/documentation/reference/api/

Reply:
> What is the benefit of adding the pinning service API to cluster when cluster already provides the required push API? Is it purely for client code generation and auth tokens?

And swapping Pinning Service providers as needed. But yeah, I don't see it is a blocker. The regular REST API can do the needed things.

Reply:
> And swapping Pinning Service providers as needed.

But that doesn't actually provide anything￼ in this scenario as the Pinning Service API doesn't allow for pushing of data, so any supporting Pinning Service Provider won't be able to accept the CAR files unless we extend the pinning services API to include the ability to push.

Not opposed to ￼specifying￼ a Pushing API that services can implement. Actually I feel like this issue is more about pushing ￼data directly and not about pinning at all so it is confusing to try and suggest that the pinning API is useful in solving this problem.